bundles/org.openhab.binding.androidtv/src/main/java/org/openhab/binding/androidtv/internal/protocol/philipstv/service/KeyPressService.java

```diff
@@ -0,0 +1,112 @@
+
+import static org.openhab.binding.androidtv.internal.protocol.philipstv.ConnectionManager.OBJECT_MAPPER;
+import static org.openhab.binding.androidtv.internal.protocol.philipstv.PhilipsTVBindingConstants.*;
+
+import java.io.IOException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.binding.androidtv.internal.protocol.philipstv.ConnectionManager;
+import org.openhab.binding.androidtv.internal.protocol.philipstv.PhilipsTVConnectionManager;
+import org.openhab.binding.androidtv.internal.protocol.philipstv.service.api.PhilipsTVService;
+import org.openhab.binding.androidtv.internal.protocol.philipstv.service.model.keypress.KeyPressDto;
+import org.openhab.core.library.types.NextPreviousType;
+import org.openhab.core.library.types.PlayPauseType;
+import org.openhab.core.library.types.RewindFastforwardType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link KeyPressService} is responsible for handling key code commands, which emulate a button
+ * press on a remote control.
+ *
+ * @author Benjamin Meyer - Initial contribution
+ * @author Ben Rosenblum - Merged into AndroidTV
+ */
+@NonNullByDefault
+public class KeyPressService implements PhilipsTVService {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final PhilipsTVConnectionManager handler;
+
+    private final ConnectionManager connectionManager;
+
+    public KeyPressService(PhilipsTVConnectionManager handler, ConnectionManager connectionManager) {
+        this.handler = handler;
+        this.connectionManager = connectionManager;
+    }
+
+    @Override
+    public void handleCommand(String channel, Command command) {
+        KeyPress keyPress = null;
+        if (isSupportedCommand(command)) {
+            // Three approaches to resolve the KEY_CODE
+            try {
+                keyPress = KeyPress.valueOf(command.toString().toUpperCase());
+            } catch (IllegalArgumentException e) {
+                try {
+                    keyPress = KeyPress.valueOf("KEY_" + command.toString().toUpperCase());
+                } catch (IllegalArgumentException e2) {
+                    try {
+                        keyPress = KeyPress.getKeyPressForValue(command.toString());
+                    } catch (IllegalArgumentException e3) {
+                        // do nothing, error message is logged later
+                    }
+                }
+            }
+
+            if (keyPress != null) {
+                try {
+                    sendKeyPress(keyPress);
+                } catch (Exception e) {
+                    if (isTvOfflineException(e)) {
+                        logger.warn("Could not execute command for key code, the TV is offline.");
```

Comment:
In respect to this, I think a message is appropriate.  In this case, the keypress has failed because the thing was actually offline but not marked correctly.  Would it not be good for the user to alert them that their command has failed because of a failure of the device?  The setting of the thing state there is just to correct the fact that it wasn't set right in the first place.

Reply:
Don;lt think a message adds anything to the case as the device end up being set to offline. I leave it up to you :-)

Reply:
It's probably more for me than anything else for troubleshooting.  I'll split the difference and move it to debug if that's OK with you?